### PR TITLE
Added BloodGlucose for read and write

### PR DESCRIPTION
**Pull request by ksetrin**

### CODE REVIEW

**Comment by aboveyunhai** [comment]
@ksetrin  Can you also include Typescript definition if you know how to do it, before I test it out tomorrow (it's midnight now).

**Comment by ksetrin** [comment]
@aboveyunhai Typescript definition for what?
I already added some here to `index.android.d.ts`
image

**Comment by aboveyunhai** [comment]
@ksetrin  My bad, I accidently take it as PR for both read/write. It's not the right time for me to review code anyway. Good Night.

**Comment by ksetrin** [comment]
@aboveyunhai added write BloodGlucose

File: android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java [hunk 1]
```diff
@@ -94,29 +100,19 @@ else if (dataReadResult.getDataSets().size() > 0) {
...
+                sample.getDouble("value"),
+                (long)sample.getDouble("date"),
+                TimeUnit.MILLISECONDS
         );
         new InsertAndVerifyDataTask(this.Dataset).execute();
 
         return true;
     }
 
-    public boolean delete(ReadableMap sample) {
```
Comment (aboveyunhai) [inline android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java:113]: I know it's probably doing nothing right now, but  it's there any reasons you remove it?

File: android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java [hunk 2]
```diff
@@ -157,24 +153,23 @@ protected Void doInBackground(Void... params) {
...
                 .build();
 
         DataSet dataSet = DataSet.create(dataSource);
-        DataPoint dataPoint = dataSet.createDataPoint().setTimeInterval(startTime, endTime, timeUnit);
+        DataPoint dataPoint = dataSet.createDataPoint();
 
-        float f1 = Float.valueOf(value.toString());
-        dataPoint = dataPoint.setFloatValues(f1);
+        dataPoint.setTimestamp(date, timeUnit);
+        dataPoint.getValue(FIELD_BLOOD_GLUCOSE_LEVEL).setFloat(Float.valueOf(value.toString()));
```
Comment (aboveyunhai) [inline android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java:172]: `Float.valueOf(value.toString())` do you mean `(float) value`

File: index.android.js [hunk 3]
```diff
@@ -525,6 +525,26 @@ class RNGoogleFit {
...
+  getBloodGlucoseSamples = async (options, callback) => {
+    const { startDate, endDate, bucketInterval, bucketUnit } = prepareInput(options);
+    const result = await googleFit.getBloodGlucoseSamples(
+      startDate,
+      endDate,
+      bucketInterval,
+      bucketUnit,
+    );
+    if (result.length > 0) {
+      return prepareResponse(result).map((i) => ({...i, value: i.value.toFixed(3)}));
```
Comment (aboveyunhai) [inline index.android.js:537]: since user will use the data to display, maybe leave the `toFixed(3)`  to the, let them figure out whatever they want, 
`toFixed()` also change the datatype from `number` to `string`

File: android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java (hunk 1 above)
Comment (ksetrin) [reply android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java:113]: Sorry, just did it out of habit, I don't like to keep not used code because it takes time and attention (I prefer keep only useful code)
I can return this method if you want

File: index.android.js (hunk 3 above)
Comment (ksetrin) [reply index.android.js:537]: not understand the part
```
 leave the toFixed(3) to the, let them figure out whatever they want,
 ```
 
 what did you mean? Need to remove `toFixed(3)` ?

File: android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java (hunk 2 above)
Comment (ksetrin) [reply android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java:172]: thanks, it's much better :)

File: index.android.js (hunk 3 above)
Comment (aboveyunhai) [reply index.android.js:537]: Irrelevant to this original question, just a slight interesting thing to share,  `prepareResponse().map()` is `O(2n)` (or O(n) technically) operation due to double iteration.
Theoretically it should have no effect, and today's device is powerful. But usually user will do their own iteration based on the data so generally we expect there will be multiple iterations happens for same dataset.  
When the amount of data is slightly large (for example, I have a 1.5 years step data while testing the app) and we want to use it to render on the screen. This side optimization may worth it, but maybe I'm worried too much.

Comment (aboveyunhai) [reply index.android.js:537]: And if user want to further modify the dataset, they can do whatever they want within one iteration along with this format.

Comment (ksetrin) [reply index.android.js:537]: @aboveyunhai make sense, already removed, but anyway I'm sure you will prettify those values from google fit after getting it

File: android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java (hunk 1 above)
Comment (aboveyunhai) [reply android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java:113]: > Sorry, just did it out of habit, I don't like to keep not used code because it takes time and attention (I prefer to keep only useful code)
> I can get back this method if you want, just let me know @aboveyunhai

Just bring it back,  because one thing I did recently is to migrate the deprecated code, these old code sometimes can give me some hints about the past and why they exist, before I remove them.

File: index.android.js (hunk 3 above)
Comment (aboveyunhai) [reply index.android.js:537]: > not understand the part
> 
> ```
>  leave the toFixed(3) to the, let them figure out whatever they want,
> ```
> 
> what did you mean? Need to remove `toFixed(3)` ?

just direct `return return prepareResponse(result);` unless I mistaken something

**Comment by ksetrin** [comment]
@aboveyunhai all done

**Comment by aboveyunhai** [comment]
> @aboveyunhai all done
LGTM.
leave me some time for me to test through it.

File: android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java [hunk 4]
```diff
@@ -157,24 +160,23 @@ protected Void doInBackground(Void... params) {
      * @param dataType DataType Fitness Data Type object
      * @param dataSourceType int Data Source Id. For example, DataSource.TYPE_RAW
      * @param value Object Values for the fitness data. They must be int or float
-     * @param startTime long Time when the fitness activity started
-     * @param endTime long Time when the fitness activity finished
+     * @param date long Time when the activity started
      * @param timeUnit TimeUnit Time unit in which period is expressed
      * @return
      */
     private DataSet createDataForRequest(DataType dataType, int dataSourceType, Double value,
```
Comment (aboveyunhai) [inline android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java:167]: `double` instead of `Double`

**Comment by aboveyunhai** [comment]
@ksetrin  
the reading part works correctly but only return partial data (`FIELD_BLOOD_GLUCOSE_LEVEL`) as far as from the testing.

DLTR: I could accept this PR since the main value works correctly, and if the feature is really important for your case. 
But I will overwrite some of codes (parts of code are still using deprecate) and add all the optional fields, enum etc. in the future. 

To give you a idea, the  `processData()` used internally is more of a general approach in the past but lacks of `options` field to handle different types of `Fintess data`. 
This is one of the annoying part of this project, if you ever had a chance to read through other files. They all have their own process data function due to `datatype`, `field datatype` difference.

The current `processData()` ignore all options fields from the blood_glucose. 

Similar issue exists for saving, all the optional fields are missing, Ideally, we may/may not include them depending on the user input. As well as creating `Enum` to allow user the pass those `int` to represent each optional field.
https://developers.google.com/fit/scenarios/write-blood-glucose-data?hl=en

https://developers.google.com/fit/datatypes/health#blood_glucose

image

**Comment by ksetrin** [comment]
> @ksetrin
> the reading part works correctly but only return partial data (`FIELD_BLOOD_GLUCOSE_LEVEL`) as far as from the testing.
> 
> TLDR: I could accept this PR since the main value works correctly, and if the feature is really important for your case.
> But I will overwrite some of codes (parts of code are still using deprecate) and add all the optional fields, enum etc. in the future.
> 
> To give you a idea, the `processData()` used internally is more of a general approach in the past but lacks of `options` field to handle different types of `Fintess data`.
> This is one of the annoying part of this project, if you ever had a chance to read through other files. They all have their own process data function due to `datatype`, `field datatype` difference.
> 
> The current `processData()` ignore all options fields from the blood_glucose.
> 
> Similar issue exists for saving, all the optional fields are missing, Ideally, we may/may not include them depending on the user input. As well as creating `Enum` to allow user the pass those `int` to represent each optional field. There are some datatype cases need to be address otherwise it will crush app.
> https://developers.google.com/fit/scenarios/write-blood-glucose-data?hl=en
> 
> https://developers.google.com/fit/datatypes/health#blood_glucose
> 
> image

We need blood glucose now in the project. It will be good to use from the library from main repo instead of fork
But I'm not Java developer, and for me it will be good to have a consultation opportunity.  

Let's merge what we have right now and improve it over time

File: android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java (hunk 4 above)
Comment (ksetrin) [reply android/src/main/java/com/reactnative/googlefit/HeartrateHistory.java]: fixed
